### ISSUE TITLE
[GR-45919] Add support for Generational ZGC on HotSpot

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#8117]
### TL;DR

Add support for [Generational ZGC (JEP 439)](https://openjdk.org/jeps/439) to the Graal compiler.

### Goals

[JDK 21 added support](https://openjdk.org/projects/jdk/21/) for [Generational ZGC](https://openjdk.org/jeps/439), which

> Improve(s) application performance by extending the Z Garbage Collector (ZGC) to maintain separate generations for young and old objects. This will allow ZGC to collect young objects — which tend to die young — more frequently.

We plan to add support for required ZGC barriers along with any relevant performance optimizations to Graal, allowing the use of Generational ZGC when Graal is used as a JIT compiler.

### Non-Goals

- Add support for (Generational) ZGC to GraalVM Native Image
- Add support for Shenandoah GC and other GCs

### Related issues

- https://github.com/oracle/graal/issues/5050

**Comment by PhantomG27249** [comment]
Hello, are there any estimates or ideas of when this  feature will be targeted/released for graal?

**Comment by tkrodriguez** [comment]
It's targeted at 24.1

**Comment by PhantomG27249** [comment]
Thank you for letting me know this news has made my day :)

**Comment by WowYouCanReallyDance** [comment]
Very much looking forward to it

**Comment by utkuaydn** [comment]
Is including (Generational) ZGC in native-image in discussions at all? If so, what is the timeframe?

**Comment by fniephaus** [comment]
> Is including (Generational) ZGC in native-image in discussions at all?

It's listed as a non-goal of this roadmap item, so no this is not about making ZGC available in native-image. We currently don't have any plans to make ZGC available in native-image.

**Comment by toinouH** [comment]
This can be moved to "done"and closed (I guess).
https://github.com/oracle/graal/commit/1e27c43a5da0bdfbababaeaa2fe8f4d9f0553dc8
Thank you very much for your work.

**Comment by fniephaus** [comment]
Correct, @toinouH: this got merged as part of https://github.com/oracle/graal/pull/9040 :)

**Comment by WowYouCanReallyDance** [comment]
Congratulations！Waiting so long.

**Comment by fniephaus** [comment]
I'm afraid support for Generational ZGC was not backported to the release branch for JDK 23, which means it will ship with JDK 24 in March 2025. However, you can already try it out today using an early access build of [Oracle GraalVM for JDK 24](https://github.com/graalvm/oracle-graalvm-ea-builds/releases?q=jdk-24&expanded=false).

**Comment by toinouH** [comment]
@fniephaus Delay can happen.
People I talked too are already very happy that we can try it and we all thank you ❤️ 

NOTE: I noticed this [benchmark](https://github.com/oracle/graal/blob/master/compiler/mx.compiler/mx_graal_benchmark.py#L49) is still disabled; I guess it's a left-over from pre-generational zgc :)

**Comment by fniephaus** [comment]
Thanks for your understanding, @toinouH! And yes, we'll enable the tests for this very soon. 🙂

**Comment by farnoy** [comment]
FYI, the changelog update in https://github.com/oracle/graal/commit/23cbacb7e4d11224743ae5279aaf5f06699bf11a#diff-893faf83d0a6d3ea59ecb706e12d7c3f2037eab0f36621c7ea0b4411b80cfdd4L7-L9 is incorrect, as it still lists it under JDK 23

**Comment by fniephaus** [comment]
Thanks for pointing this out, @farnoy. We'll get that fixed.

**Comment by ShalokShalom** [comment]
Just as a heads up: Its not yet documented

https://www.graalvm.org/jdk24/reference-manual/native-image/optimizations-and-performance/MemoryManagement/

**Comment by toinouH** [comment]
@ShalokShalom No, ZGC isn't supported by Native image; only by the JIT for now.

**Comment by ShalokShalom** [comment]
OK, thanks. It seems like there is no documentation on garbage collection with the JIT runtime? Is this, because its the same as with OpenJDK?
